### PR TITLE
Jbang setup

### PR DESCRIPTION
**Pull request by quintesse**
It works fine on Linux.

But it's the Windows implementation I'm worried about, it's more complex because of the fact that we can't overwrite any files that are currently in use. So I had to come up with a ping-pong mechanism of handing execution over from one script to another. I haven't been able to test any of that because I don't have access to a Windows machine right now. So I will need help validating what I implemented (especially the powershell version is most likely incorrect).

### CODE REVIEW

File: src/main/java/dev/jbang/Util.java [hunk 1]
```diff
@@ -733,33 +733,31 @@ public static String runCommand(String... cmd) {
...
 	}
 
-	public static boolean deleteFolder(Path folder, boolean quiet) {
-		boolean result[] = new boolean[] { true };
-		if (Files.isDirectory(folder)) {
-			Util.verboseMsg("Deleting " + folder);
-			try {
-				Files	.walk(folder)
+	public static boolean deletePath(Path path, boolean quiet) {
+		Exception err[] = new Exception[] { null };
```
Comment (tivrfoa) [inline src/main/java/dev/jbang/Util.java:737]: Does it need to be an array?

**Comment by maxandersen** [comment]
I can try on windows - hopefully later tonight. 

Btw. Did you do this all from scratch or copied others install approach ?

File: src/main/java/dev/jbang/Util.java (hunk 1 above)
Comment (quintesse) [reply src/main/java/dev/jbang/Util.java:737]: Yes, unfortunately this is the trick one has to use when updating a var from within a lambda block.

**Comment by quintesse** [comment]
> Btw. Did you do this all from scratch or copied others install approach ?

From scratch, it just came out this way in one go, but it might be good to see if there are others doing a much better job, just preferred the coding over doing research :wink: 

Btw, I just updated it with a `--with-java` option which will also add Jbang's JDK to the user's environment (PATH and JAVA_HOME).

Also, running the setup multiple times won't update already added configuration on Linux/Mac, that's still a TODO.

File: src/main/java/dev/jbang/Util.java (hunk 1 above)
Comment (tivrfoa) [reply src/main/java/dev/jbang/Util.java:737]: Interesting hack to know. ty =D

File: src/main/java/dev/jbang/cli/Setup.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.nio.file.*;
+import java.security.CodeSource;
+
+import dev.jbang.*;
+
+import picocli.CommandLine;
+
+@CommandLine.Command(name = "setup", description = "Make Jbang available for the user, either in the current session or permanently")
```
Comment (maxandersen) [inline src/main/java/dev/jbang/cli/Setup.java:13]:
```suggestion
@CommandLine.Command(name = "setup", description = "Make jbang available for the user, either in the current session or permanently")
```

File: src/main/java/dev/jbang/cli/Setup.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+public class Setup extends BaseCommand {
+	private static final String jbangUrl = "https://github.com/jbangdev/jbang/releases/latest/download/jbang.zip";
+
+	@CommandLine.Option(names = {
+			"--with-java" }, description = "Add Jbang's Java to the user's environment as well")
+	boolean withJava;
+
+	@CommandLine.Option(names = {
+			"--fresh" }, description = "Force re-download and re-install of Jbang")
```
Comment (maxandersen) [inline src/main/java/dev/jbang/cli/Setup.java:22]:
```suggestion
			"--fresh" }, description = "Force re-download and re-install of jbang")
```

File: src/main/java/dev/jbang/cli/Setup.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+				Util.deletePath(toDir, true);
+				copyJbangFiles(fromDir, toDir);
+				createUpdateScript(toDir);
+			} else {
+				copyJbangFiles(fromDir, binDir);
+			}
+		}
+		String cmd = "";
+		// Permanently add Jbang's bin folder to the user's PATH
+		Util.infoMsg("Adding Jbang to PATH...");
```
Comment (maxandersen) [inline src/main/java/dev/jbang/cli/Setup.java:64]: log how/where it does it - at least in verbose mode.

File: src/main/java/dev/jbang/cli/Setup.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	private static void changeScript(Path binDir, Path javaHome, Path bashFile) throws IOException {
+		// Detect if Jbang has already been set up before
+		boolean jbangFound = Files	.lines(bashFile)
+									.anyMatch(ln -> ln.trim().startsWith("#") && ln.toLowerCase().contains("jbang"));
+		if (!jbangFound) {
+			// Add lines to add Jbang to PATH
+			String lines = "\n# Add Jbang to environment\n" +
+					"alias j!=jbang\n";
```
Comment (maxandersen) [inline src/main/java/dev/jbang/cli/Setup.java:112]: That was fast :)

**Comment by maxandersen** [comment]
in powershell i get this (after nuking path for jbang refs with `rundll32 sysdm.cpl,EditEnvironmentVariables`) and run build/install/jbang setup:
```
PS Z:\code\personal\jbang\build\install> .\jbang\bin\jbang setup
[jbang] Adding Jbang to PATH...
& : The term '[Environment]::SetEnvironmentVariable' is not recognized as the name of a cmdlet, function, script file,
or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and
try again.
At line:1 char:3
+ & [Environment]::SetEnvironmentVariable('Path', 'C:\Users\max\.jbang\ ...
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: ([Environment]::SetEnvironmentVariable:String) [], CommandNotFoundExcept
   ion
    + FullyQualifiedErrorId : CommandNotFoundException
```

**Comment by maxandersen** [comment]
`[System.Environment]::SetEnvironmentVariable(` seem to be the right thing that works on powershell.

File: src/main/java/dev/jbang/cli/Setup.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+			}
+			String env = "[Environment]::SetEnvironmentVariable('Path', '" + newPath + "' + " +
+					"[Environment]::GetEnvironmentVariable('Path', [EnvironmentVariableTarget]::User), " +
+					"[EnvironmentVariableTarget]::User)";
+			if (withJava) {
+				env += " ; [Environment]::SetEnvironmentVariable('JAVA_HOME', '" + jdkHome + "', " +
+						"[EnvironmentVariableTarget]::User)";
+			}
+			if (Util.isUsingPowerShell()) {
+				cmd += env + " ; ";
```
Comment (maxandersen) [inline src/main/java/dev/jbang/cli/Setup.java:79]:
```suggestion
				cmd += env + " ; ";
				cmd = "{ " + cmd + " }";
```

Comment (maxandersen) [reply src/main/java/dev/jbang/cli/Setup.java:79]: fixes the error on powershell

**Pull request merged**